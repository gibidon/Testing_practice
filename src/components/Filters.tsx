import { useState } from "react"
import { useDispatch, useSelector } from "react-redux"
import { RootState } from "src/store/configureStore"
import { toggleOnlyActive, updateFilterParams } from "src/store/filterSlice"

interface Filters {
    onlyFiltered:boolean
    
}

const initialFilters:Filters = {
    onlyFiltered:false
}

export const Filters = () => {
    

    return (<div>
        <Switch/>
    </div>)
    
}

const Switch = () => {
    const dispatch = useDispatch()
    const onlyActive = useSelector((state:RootState) =>state.filters.onlyActive)

    return <div>
        <button data-testid='onlyActiveTasksBtn' className="onlyActiveTasksBtn" onClick={() => dispatch(toggleOnlyActive())}>{onlyActive ? 'Show all' : 'Show only active'}</button>
        {/* <button onClick={() => dispatch(updateFilterParams({key:"onlyActive",value:!onlyActive}))}>{onlyActive ? 'Show all' : 'Show only active'}</button> */}
        {/* <button onClick={() => dispatch(updateFilterParams({key:"genre",value:genre ==='male' ? 'female' : 'male'}))}>{genre === 'male' ? 'female' : 'male'}</button> */}
    </div>
}