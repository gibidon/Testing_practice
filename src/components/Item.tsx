import { validateHeaderMax, validateHeaderMin,HEADER_LENGTH_ERROR_MESSAGE } from "src/utils/helpers";
import { DeleteButton } from "./DeleteButton";

type Props = Task & {
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const Item = (props: Props) => {
  // const disabled = validateHeaderMax(props.header) || validateHeaderMin(props.header)
  const disabled = validateHeaderMax(props.header) && validateHeaderMin(props.header)

  return (
    <li className="item-wrapper">
      <input
        type="checkbox"
        id={props.id}
        defaultChecked={props.done}
        onChange={() => props.onToggle(props.id)}
      />
      <label htmlFor={props.id} onClick={() => props.onToggle(props.id)}>
        {/* {props.done ? <s>{props.header}</s> : props.header} */}
        {disabled ? props.header :HEADER_LENGTH_ERROR_MESSAGE}
      </label>
      <DeleteButton
        disabled={!props.done}
        onClick={() => props.onDelete(props.id)}
      />
    </li>
  );
};
