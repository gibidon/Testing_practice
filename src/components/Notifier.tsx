import { useEffect } from "react";

type Props = {
    task: string;
    open: boolean;
    onClose: () => void;
}

export const Notifier = ({ task, open, onClose }: Props) => {
    useEffect(() => {
        let timerId: NodeJS.Timeout;
        console.log('applying setTimeout in useEffect')
        if (open) {
            timerId = setTimeout(() => {
                onClose();
            }, 2000);
        }

        return () => {
            console.log('applying setTimeout in useEffect')

            clearTimeout(timerId)};
    }, [open]);

    if (!open) {
        return null;
    }

    return(
        <div className="blackout">
            <div className="notifier-wrapper" data-testid='notifier'>
                {task}
            </div>
        </div>
    );
}