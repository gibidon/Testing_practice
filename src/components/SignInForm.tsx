import { ChangeEvent, useState,useRef, FormEvent, FormEventHandler } from "react"

const initialInputState = {
    name:'',
    surname:'',
    age:0
}

export const SignInForm = () => {
    const [inputs,setInputs] = useState(initialInputState)
    const formRef = useRef<HTMLFormElement>(null)
    const inputNameRef = useRef<HTMLInputElement>(null)
    const handleChange:FormEventHandler = (e:ChangeEvent<HTMLInputElement>) => {
        setInputs((prevState) => ({...prevState,[e.target.name]:e.target.value
        }))
    }

    const handleSubmit = (e:FormEvent<HTMLFormElement>) => {
        e.preventDefault()
        console.log('inputs submitted: ',inputs )
        
        formRef.current?.reset()
    }

    const handleReset = () => {
        setInputs(initialInputState)
    }

    const handleBegin = () => {
        inputNameRef.current?.focus()
    }
    return (
        <>
        <p><button onClick={handleBegin}>Fill out the form</button></p>
         <form 
            ref={formRef}
            onChange={handleChange}
            onSubmit={handleSubmit}
            onReset={handleReset}
        >
            <input  type='text' name='name' ref={inputNameRef}/>
            <input  type='text' name='surname'/>
            <input  type='number' name='age'/>
            <p>
                <button type='submit'>Submit</button>
                <button type='reset'>Reset</button>
            </p>
        </form>
        </>
       
    )
}