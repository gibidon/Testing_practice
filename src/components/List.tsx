import { Item } from "./Item";

type Props = {
  items: Task[];
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const List = ({ items, onDelete, onToggle }: Props) => {
  const itemsToRender = items.length <= 10 ? items : items.slice(0,9) 

  return (
  <ul className="task-list tasks">
    {/* {items.map((item) => ( */}
    {itemsToRender.map((item) => (
      <Item
        {...item}
        key={item.id}
        onDelete={onDelete}
        onToggle={onToggle} />
    ))}
  </ul>
)
}