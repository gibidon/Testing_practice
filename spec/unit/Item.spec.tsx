import { render,screen,prettyDOM } from "@testing-library/react";
import { Item } from "src/components/Item";
import ue from '@testing-library/user-event'
import { HEADER_LENGTH_ERROR_MESSAGE } from "src/utils/helpers";

const LONG_STRING = 'string that is too long and exceeds 32 symbols for testing purposes'
const NORMAL_STRING = 'some normal string'
const EMPTY_STRING = ''

describe('Элемент списка задач', () => {
    const userEvent = ue.setup({advanceTimers:jest.advanceTimersByTime})

    test('название не должно быть больше 32 символов',() => {
        const onDelete = jest.fn()
        const onToggle = jest.fn()
        
        render(<Item header={LONG_STRING} id="1" done={true} onDelete={onDelete} onToggle={onToggle}/>)
        
        const item = screen.getByText(HEADER_LENGTH_ERROR_MESSAGE)

        expect(item.textContent).toBe(HEADER_LENGTH_ERROR_MESSAGE)
    });
    test('название не должно быть пустым',() => {
        const onDelete = jest.fn()
        const onToggle = jest.fn()
        
        render(<Item header={EMPTY_STRING} id="1" done={false} onDelete={onDelete} onToggle={onToggle}/>)
        
        const item = screen.getByText(HEADER_LENGTH_ERROR_MESSAGE)

        expect(item.textContent).toBe(HEADER_LENGTH_ERROR_MESSAGE)
    });

    test('нельзя удалять невыполненные задачи',async() => {
        const onDelete = jest.fn()
        const onToggle = jest.fn()

        const view = render(<Item header={NORMAL_STRING} id="1" done={false} onToggle={onToggle} onDelete={onDelete}/>)
       
    
        const deleteBtn = view.getByRole('button')
        console.log('deleteBtn: ',prettyDOM(deleteBtn))
        
        expect(deleteBtn).toBeDisabled()
    });

    it('заполняется корректно-через diffSnapshot',() => {
        const onToggle = jest.fn()
        const onDelete = jest.fn()

        const {rerender,asFragment} = render(<Item header={NORMAL_STRING} id='6' done={true} onToggle={onToggle} onDelete={onDelete}/>)
        const firstRender = asFragment()

        rerender(<Item header={NORMAL_STRING + 'new'} id='7' done={true} onToggle={onToggle} onDelete={onDelete}/>)
        const secondRender = asFragment()

        expect(firstRender).toMatchDiffSnapshot(secondRender)
    })
    it('заполняется корректно-через snapshot',() => {
        const onToggle = jest.fn()
        const onDelete = jest.fn()

        const {asFragment} = render(<Item header={NORMAL_STRING} id='6' done={true} onToggle={onToggle} onDelete={onDelete}/>)
        const firstRender = asFragment()

        expect(firstRender).toMatchSnapshot()
    })
   
    
});