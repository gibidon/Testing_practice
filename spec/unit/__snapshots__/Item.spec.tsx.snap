// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Элемент списка задач заполняется 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -2,17 +2,17 @@
    <li
      class="item-wrapper"
    >
      <input
        checked=""
-       id="6"
+       id="7"
        type="checkbox"
      />
      <label
-       for="6"
+       for="7"
      >
-       some normal string
+       some normal stringnew
      </label>
      <button
        class="button button-with-icon"
        data-alt="Удалить"
      >"
`;

exports[`Элемент списка задач заполняется корректно-через diffSnapshot 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -2,17 +2,17 @@
    <li
      class="item-wrapper"
    >
      <input
        checked=""
-       id="6"
+       id="7"
        type="checkbox"
      />
      <label
-       for="6"
+       for="7"
      >
-       some normal string
+       some normal stringnew
      </label>
      <button
        class="button button-with-icon"
        data-alt="Удалить"
      >"
`;

exports[`Элемент списка задач заполняется корректно-через snapshot 1`] = `
<DocumentFragment>
  <li
    class="item-wrapper"
  >
    <input
      checked=""
      id="6"
      type="checkbox"
    />
    <label
      for="6"
    >
      some normal string
    </label>
    <button
      class="button button-with-icon"
      data-alt="Удалить"
    >
      <img
        alt="Удалить"
        src="[object Object]"
      />
    </button>
  </li>
</DocumentFragment>
`;
